### ISSUE TITLE
Interacting with privateGPT (specifically for RAG)

### ISSUE THREAD

**Issue by Aquan1412** [karthink/gptel#305]
Hello, I'm looking for a way to use privateGPT as a backend for gptel, in order to use its simple RAG pipeline. 
Specifically, I'm looking for a way to provide additional keywords to the backend (such as "use_context" and "include_sources").

I can generally use privateGPT as an "openai"-like backend with the following configuration:

```
(gptel-make-openai "privateGPT"
  :protocol "http"
  :host "localhost:8001"
  :models '("private-gpt"))
```

I tried simply adding the keyword to the configuration:

```
(gptel-make-openai "privateGPT"
  :protocol "http"
  :host "localhost:8001"
  :use_context t
  :models '("private-gpt"))
```

However, if I do that, I get the following error message:

>   Debugger entered--Lisp error: (error "Keyword argument :use_context not one of (:curl-ar...")
>   signal(error ("Keyword argument :use_context not one of (:curl-ar..."))
>   error("Keyword argument %s not one of (:curl-args :models..." :use_context)
>   gptel-make-openai("privateGPT-Context" :protocol "http" :host "localhost:8001" :use_context t :models ("private-gpt"))
>   (progn (gptel-make-openai "privateGPT-Context" :protocol "http" :host "localhost:8001" :use_context t :models '("private-gpt")))
>   eval((progn (gptel-make-openai "privateGPT-Context" :protocol "http" :host "localhost:8001" :use_context t :models '("private-gpt"))) t)
>   elisp--eval-last-sexp(nil)
>   eval-last-sexp(nil)
>   funcall-interactively(eval-last-sexp nil)
>   call-interactively(eval-last-sexp nil nil)
>   command-execute(eval-last-sexp)

So, is there a way to get this to work? Or am I approaching it completely wrong?

**Comment by karthink** [comment]
You're going to have to provide more context for me to understand what you're looking for.  What do you expect `use_context` and `include_sources` to do?

**Comment by Aquan1412** [comment]
I want to use privateGPT to generate responses based on embeddings I previously generated by "ingesting" several PDFs (in privateGPT lingo). According to the [API reference](https://docs.privategpt.dev/api-reference/api-reference/contextual-completions/chat-completion-v-1-chat-completions-post-stream), if I set `use_context = true`, it should use the embeddings to generate responses based on the ingested papers. Additionally, if I set `include_sources = true`, it should include the source chunks, on which the generated answers are based.

**Comment by karthink** [comment]
> I believe inserting:
> ``` elisp
>   :use_context t
>   :include_sources t
> ```
> after line 103 will do what you want but is likely to break other things. This change will add those key/value pairs to the data sent to the OpenAI API through curl.

This is correct.  I'll need to find some way of adding this via the configuration.

**Comment by karthink** [comment]
> Maybe I'll try and create a separate gptel-make-privateGPT based on gptel-make-openai. Now that I roughly know where to look, i shouldn't be too hard.

You'll need to write a new struct type `gptel-privategpt` that inherits from `gptel-openai`, and three `cl-defmethod`s that specialize on the backend-type `gptel-privategpt`: `gptel--request-data`, `gptel-curl--parse-stream` and `gptel--parse-response`.  We can add it to gptel afterwards.  I can help if you have questions.

**Comment by Aquan1412** [comment]
> > Maybe I'll try and create a separate gptel-make-privateGPT based on gptel-make-openai. Now that I roughly know where to look, i shouldn't be too hard.
> 
> You'll need to write a new struct type `gptel-privategpt` that inherits from `gptel-openai`, and three `cl-defmethod`s that specialize on the backend-type `gptel-privategpt`: `gptel--request-data`, `gptel-curl--parse-stream` and `gptel--parse-response`. We can add it to gptel afterwards. I can help if you have questions.

Great, thanks for the information! I'll see how far I get, and if I encounter any big problems, I'll come back to you.

**Comment by karthink** [comment]
Thanks!  Would you like to raise a PR?  I can review the code and we can add it to gptel.

**Comment by Aquan1412** [comment]
Sure, I just raised the PR. Let me know if there any further changes necessary.

**Comment by karthink** [comment]
PrivateGPT support has been added.
